### PR TITLE
Integer labels

### PR DESCRIPTION
**Pull request by DeNeutoy**
This PR does a few things to the data API:

- Renames `TagField` to `SequenceLabelField`.
- Makes `SequenceLabelField` and `LabelField` return integer labels.
- Allows `SequenceLabelField` to accept either integer labels or string labels - in the integer case, the field does not get indexed.
- Removes the checks for 3D tags in the various models.
- Renames the `.tag()` method of `SequenceLabelField` to `labels()`
- Changes the default `namespace` in `SequenceLabelField` to be `labels` not `tags`.
- Removes all reference to `num_tags` in all `Fields`.

Instead of adding a `SequenceFeatureField`, making `SequenceLabelField` able to ignore indexing directly gives us all functionality we need to use it for feature representation as well as labels. I could re-name this to `ConditionalTextField` or something? Not sure what the best name is here, as it doesn't make a distinction between labels vs inputs.

One question - I changed the `LabelField` to take a flag `skip_indexing`, if the user wants to skip indexing (previously this was managed by passing the total number of tags, but this is now not required). In the new `SequenceLabelField` I have automatically assumed that if the labels are integers, we should skip indexing. Which do you prefer?

### CODE REVIEW

File: allennlp/data/fields/label_field.py [hunk 1]
```diff
@@ -16,43 +17,42 @@ class LabelField(Field[numpy.ndarray]):
...
-    of unique labels in your data.
+    This field will get converted into an integer index representing the class label.
 
     Parameters
     ----------
     label : ``Union[str, int]``
     label_namespace : ``str``, optional (default="labels")
         The namespace to use for converting label strings into integers.  We map label strings to
-        integers for you (e.g., "entailment" and "contradiction" get converted to one-hot vectors),
+        integers for you (e.g., "entailment" and "contradiction" get converted 0, 1, ...),
```
Comment (matt-gardner) [inline allennlp/data/fields/label_field.py]: "converted to"

File: allennlp/data/fields/sequence_label_field.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    Parameters
+    ----------
+    labels : ``List[str]``
+        A sequence of categorical labels, encoded as strings.  These could be POS tags like [NN,
+        JJ, ...], BIO tags like [B-PERS, I-PERS, O, O, ...], or any other categorical tag sequence.
+    sequence_field : ``SequenceField``
+        A field containing the sequence that this ``SequenceLabelField`` is labeling.  Most often, this is a
+        ``TextField``, for tagging individual tokens in a sentence.
+    label_namespace : ``str``, optional (default='tags')
```
Comment (matt-gardner) [inline allennlp/data/fields/sequence_label_field.py]: `s/tags/labels/`

File: allennlp/data/fields/sequence_label_field.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    A ``SequenceLabelField`` assigns a categorical label to each element in a :class:`SequenceField`.
+    Because it's a labeling of some other field, we take that field as input here, and we use it to
+    determine our padding and other things.
+
+    This field will get converted into a list of integer class ids, representing the correct class
+    for each element in the sequence.
+
+    Parameters
+    ----------
+    labels : ``List[str]``
```
Comment (matt-gardner) [inline allennlp/data/fields/sequence_label_field.py]: Update type here, and note that if the input is all ints, we won't use a vocab to convert it.

File: allennlp/data/fields/sequence_label_field.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                                     "don't match: %d and %d" % (len(labels), sequence_field.sequence_length()))
+
+        if all([isinstance(x, int) for x in labels]):
+            self._indexed_labels = labels
+
+    @overrides
+    def count_vocab_items(self, counter: Dict[str, Dict[str, int]]):
+        if self._indexed_labels is None:
+            for label in self._labels:
+                counter[self._label_namespace][label] += 1  # type: ignore
```
Comment (matt-gardner) [inline allennlp/data/fields/sequence_label_field.py]: Why do we need these `type: ignore` statements?  They weren't there in `TagField`.

Comment (matt-gardner) [inline allennlp/data/fields/sequence_label_field.py]: Oh, because of the `Union`?

File: allennlp/models/bidaf.py [hunk 5]
```diff
@@ -214,15 +214,10 @@ def forward(self,  # type: ignore
...
-            # Negative log likelihood criterion takes integer labels, not one hot.
-            if span_start.dim() == 2:
-                _, span_start = span_start.max(-1)
-            loss = nll_loss(util.masked_log_softmax(span_start_logits, passage_mask), span_start.view(-1))
-            self._span_start_accuracy(span_start_logits, span_start)
-            if span_end.dim() == 2:
-                _, span_end = span_end.max(-1)
-            loss += nll_loss(util.masked_log_softmax(span_end_logits, passage_mask), span_end.view(-1))
-            self._span_end_accuracy(span_end_logits, span_end)
+            loss = nll_loss(util.masked_log_softmax(span_start_logits, passage_mask), span_start.squeeze(-1))
```
Comment (matt-gardner) [inline allennlp/models/bidaf.py]: Seems odd to have to do a squeeze here still.  Is there a reasonable way to do the squeeze earlier in the data pipeline?

File: allennlp/service/predictors/semantic_role_labeler.py [hunk 6]
```diff
@@ -25,7 +25,9 @@ def predict_json(self, inputs: JsonDict) -> JsonDict:
         spacy_doc = self.nlp(sentence)
         for i, word in enumerate(spacy_doc):
             if word.pos_ == "VERB":
-                verb_indicator = IndexField(i, text)
+                this_verb_indicator = [0 for _ in tokens]
```
Comment (matt-gardner) [inline allennlp/service/predictors/semantic_role_labeler.py]: `verb_labels`? `verb_indicator_array`?

File: tests/data/fields/label_field_test.py [hunk 7]
```diff
@@ -1,17 +1,19 @@
...
 from allennlp.data.fields import LabelField
 from allennlp.common.testing import AllenNlpTestCase
+from allennlp.common.checks import ConfigurationError
 
 
 class TestLabelField(AllenNlpTestCase):
 
-    def test_pad_returns_one_hot_array(self):
-        label = LabelField(5, num_labels=10)
+    def test_pad_returns_integer_array(self):
```
Comment (matt-gardner) [inline tests/data/fields/label_field_test.py]: Looks like we missed renaming the tests themselves when we switched from `.pad()` to `.as_array()`.  Can you fix this one, at least?

File: tests/data/fields/sequence_label_field_test.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+        tags = ["B", "I", "O", "O", "O"]
+        sequence_label_field = SequenceLabelField(tags, self.text, label_namespace="*labels")
+        sequence_label_field.index(vocab)
+
+        # pylint: disable=protected-access
+        assert sequence_label_field._indexed_labels == [b_index, i_index, o_index, o_index, o_index]
+        # pylint: enable=protected-access
+
+    def test_pad_produces_integer_targets(self):
```
Comment (matt-gardner) [inline tests/data/fields/sequence_label_field_test.py]: Here too - `test_as_array_produces...`

File: allennlp/data/fields/sequence_label_field.py (hunk 4 above)
Comment (DeNeutoy) [inline allennlp/data/fields/sequence_label_field.py]: Yes
